README.md
# i3weather-dwd

i3status plugin for displaying weather information from the German Meteorological
Service (Deutscher Wetterdienst). 

The DWD (Deutscher Wetterdienst) publishes most of its data on its
[open data platform](https://www.dwd.de/DE/leistungen/opendata/opendata.html).

In fact `i3weather-dwd` isn't a plugin, but it uses the `read_file` module from i3status.
`i3weather-dwd` creates and updates the file `/tmp/i3weather-dwd` regularly and the
`read_file` module will read the file on change.

`i3weather-dwd` could run as a **daemon** or may be invoked by **cron**.

The Information is displayed in the following manner:

    10:00 (UTC) T: 6,5 °C  H: 78 %  W: 28 km/h (bedeckt)

or in english:

    10:00 (UTC) T: 6,5 °C  H: 78 %  W: 28 km/h (overcast)




T: temperature in degree celsius (2 meter above surface). 

H: relative air humidity.

W: average wind speed last hour.

(meterological description)


## installation

Clone the repository an make sure you have `python3 >= 3.5` and `i3status >= 2.13` installed.


    i3status -v
    i3status 2.13 © 2008 Michael Stapelberg and contributors
    
    python3 --version
    Python 3.8.2

    git clone --depth=1 https://github.com/meetunix/i3weather-dwd.git


Copy the executable (python script) to a location of your choice.


    cp i3weather-dwd/i3weather-dwd.py /usr/local/bin/


## configuration

### weather station lookup

Every weather station has a unique id. You need to find out this number to use
`i3weather-dwd`. A weather station located in germany is updated **hourly**. You will
find the same information on their
[website](https://www.dwd.de/DE/wetter/wetterundklima_vorort/mecklenburg-vorpommern/warnemuende/_node.html).


You have the choice beetween a live lookup using the `station-lookup.py` script, or
searching a generated snapshot of the weather stations in [stations.md](stations.md).


Using the live lookup is very easy. If there is no station in your town, maybe there is
one in the neighbouring town.

    i3weather-dwd/stationLookup.py rostock
    no station found
    
    i3weather-dwd/stationLookup.py warnemünde
    10170 <--> warnemuende

The station id for the weather station located in Warnemünde is 10170.


### running i3weather-dwd once or with cron 

If you want to start the `i3weather-dwd` a single time, just run it.

    /usr/local/bin/i3weather-dwd.py -s STATION-ID


Maybe you want to use cron for regulary updates

    */20 * * * *    /usr/local/bin/i3weather-dwd.py -s STATION-ID

That means: every 20 minutes an update is performed.



### updating your i3status.conf

Simply add the following lines to your config (mostly `~/.i3status.conf`) and reload i3.

    order += "read_file weather"

    ...


    read_file weather {
        format = %content
        path = "/tmp/i3weather-dwd"
    }


### command line options
    
    # mandatory

    -s --station    STRING      Station-ID

    # optional

    -d --daemon     FLAG        run as daemon (experimental)
    -p --pid        STRING      alternate path to the pid

    -g --german     FLAG        meterological description is displayed in german

### running as a daemon (experimental)

You are able to run `i3weather-dwd` as a daemon. The prefered way is the use
of **systemd**. But you can also start it like this:

    /usr/local/bin/i3weather-dwd.py -s STATION-ID -d
    

#### using systemd

Copy the following [file](contrib/i3weather-dwd.service) to
`/etc/systemd/system/i3weather-dwd.service` and change the user and group names under
which the service has to be started.

    [Unit]
    Description=i3weather-dwd

    [Service]
    User=USERNAME
    Group=GROUPNAME
    Type=forking
    ExecStart=/usr/local/bin/i3weather-dwd.py -s 10170 -d
    PIDFile=/tmp/i3weather-dwd.pid
    TimeoutStartSec=5

    [Install]
    WantedBy=multi-user.target

Then start and enable the service.


    sudo systemctl daemon-reload
    
    sudo systemctl start i3weather-dwd.service
    
    sudo systemctl enable i3weather-dwd.service

Have fun!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.3] - 2020-04-18

### Fixed

- daemon does not crash if internet connection is not available on startup



## [0.1.2] - 2020-04-05

### Fixed

- time of measurement is correctly displayed 



## [0.1.1] - 2020-04-03

### Added
- systemd service-file

### Changed
- behaviour of the daemonization

### Fixed



## [0.1.0] - 2020-04-02

### Added
- parameter for alternate pid creation 

### Changed
- syntax is now more PEP-8 compatible 

### Fixed
- station lookup with german umlauts and the sharp s is now possible


station-lookup.py
#!/usr/bin/env python3
""" station-lookup - searches the DWD-weather-station catalogue inplace

Copyright © 2020 Martin Steinbach

See file LICENSE for license information
"""

import sys
import re
import urllib.request

# Quelle: Deutscher Wetterdienst
ALL_STATION_URL = 'https://www.dwd.de/DE/leistungen/met_verfahren_mosmix/mosmix_stationskatalog.cfg?view=nasPublication&nn=16102'
AVAILABLE_STATIONS_URL = 'https://opendata.dwd.de/weather/weather_reports/poi/'

def get_file_from_url(url):
    response = urllib.request.urlopen(url)
    #print("returned status: {} - file size: {}".
    #    format(response.status,response.getheader("Content-Length")))
    if response.status != 200:
        return None
    else:
        return response.read()
    
def get_testfile(path):
    fb = open(path,"rb")
    f = fb.read()
    fb.close()
    return f

def build_stations_dict(stat_list, avail_stat_list):
    stations_dict = {}
    for l in stat_list:
        if re.match('^9',l):
            l = l[12:]
            sId = l.split(' ')[0]
            if sId in avail_stat_list:
                place = l[11:32].strip().lower() 
                stations_dict.update({sId : place})
    return stations_dict

def build_available_stations_list(ll):
    available_stations = []
    for l in ll:
        if re.match('^<a href',l):
            l = l.split("\"")
            l = l[1].split("-")[0].replace("_"," ").strip()
            available_stations.append(l)
    return available_stations

def convert_umlauts (search_input):
    
    umlauts = {
            b'\xc3\xa4': b'ae',  # U+00E4
            b'\xc3\xb6': b'oe',  # U+00F6
            b'\xc3\xbc': b'ue',  # U+00FC
            b'\xc3\x84': b'Ae',  # U+00C4
            b'\xc3\x96': b'Oe',  # U+00D6
            b'\xc3\x9c': b'Ue',  # U+00DC
            b'\xc3\x9f': b'ss',  # U+00DF
            }

    search_utf8 = search_input.encode('utf-8')

    for umlaut in umlauts.keys():
        search_utf8 = search_utf8.replace(umlaut,umlauts[umlaut])
   
    return search_utf8.decode()

def main(search):
    
    avail_stations = get_file_from_url(AVAILABLE_STATIONS_URL)
    avail_stations = avail_stations.decode('utf_8').splitlines()
    avail_stations = build_available_stations_list(avail_stations)

    stations = get_file_from_url(ALL_STATION_URL)
    #stations = get_testfile('mosmix_stationskatalog.cfg') 
    stations = stations.decode("iso8859-1").splitlines()
    stations_dict = build_stations_dict(stations, avail_stations)

    search = search.strip().lower()
    search = convert_umlauts(search)
    pattern = '.*'+search+'.*'
    pattern = re.compile(pattern)
    matches = 0
    for key, place in stations_dict.items():
        if re.match(pattern,place):
            matches += 1
            print("{:5} <--> {}".format(key,place))

    if matches == 0:
        print("no station found")


if __name__ == "__main__":
    main(sys.argv[1])



i3weather-dwd.py
#!/usr/bin/env python3

""" i3weather-dwd

i3status plugin for displaying weather information from the German Meteorological
Service (Deutscher Wetterdienst) without external dependencies.

Copyright © 2020 Martin Steinbach

See file LICENSE for license information
"""

import os
import sys
import csv
import argparse
import signal
import urllib.request
import resource as rsc
from time import sleep
from pathlib import Path

# Quelle: Deutscher Wetterdienst
DWD_POI_URL = "https://opendata.dwd.de/weather/weather_reports/poi/"

# FILE_PATH: Path to the file where the weather information is stored.
FILE_PATH = "/tmp/i3weather-dwd"

# Default path to pid file if no argument is passed to --pid:
PID_DEFAULT = "/tmp/i3weather-dwd.pid"

#https://www.dwd.de/DE/leistungen/opendata/help/schluessel_datenformate/bufr/
#poi_present_weather_zuordnung_pdf.pdf?__blob=publicationFile&v=2
# code -> (shortDescr, detailDescr, englishDescr)
PRESENT_WEATHER_CODES ={  
    "1" : ("wolkenlos","wolkenlos","clear"),
    "2" : ("heiter","heiter","bright"),
    "3" : ("bewölkt","bewoelkt","cloudy"),
    "4" : ("bedeckt","bedeckt","overcast"),
    "5" : ("Nebel","Nebel","fog"),
    "6" : ("Nebel","gefrierender Nebel","fog"),
    "7" : ("Regen","leichter Regen","rain (sprinkle)"),
    "8" : ("Regen","Regen","rain"),
    "9" : ("Regen","kräftiger Regen","heavy rain"),
    "10" : ("Schneeregen","gefrierender Regen","sleet"),
    "11" : ("Schneeregen","kraeftiger gefrierender Regen","sleet"),
    "12" : ("Schneeregen","Schneeregen","sleet"),
    "13" : ("Schneeregen","kraeftiger Schneeregen","sleet"),
    "14" : ("Schneefall","leichter Schneefall","snowfall"),
    "15" : ("Schneefall","Schneefall","snowfall"),
    "16" : ("Schneefall","kraeftiger Schneefall","snowfall"),
    "17" : ("Schneefall","Eiskoerner","hail"),
    "18" : ("Regenschauer","Regenschauer","rain shower"),
    "19" : ("Regenschauer","kraeftiger Regenschauer","rain shower"),
    "20" : ("Schneeregenschauer","Schneeregenschauer","heavy sleet"),
    "21" : ("Schneeschauer","kraeftiger Schneeregenschauer","heavy sleet"),
    "22" : ("Schneeschauer","Schneeschauer","snow shower"),
    "23" : ("Schneeschauer","kraeftiger Schneeschauer","heavy snow shower"),
    "24" : ("Schneeschauer","Graupelschauer","hail pellet shower"),
    "25" : ("Schneeschauer","kraeftiger Graupelschauer","heavy hail pellet shower"),
    "26" : ("Gewitter","Gewitter ohne Niederschlag","thunderstorm wo precipitation"),
    "27" : ("Gewitter","Gewitter","thunderstorm"),
    "28" : ("Gewitter","kraeftiges Gewitter","heavy thunderstorm"),
    "29" : ("Gewitter","Gewitter mit Hagel","thunderstorm with hail"),
    "30" : ("Gewitter","kraeftiges Gewitter mit Hagel","heavy thunderstorm with hail"),
    "31" : ("Sturm","Boen","storm")}

def daemonize():

    WORKDIR = "/tmp"
    UMASK = 0o033
    REDIRECT_FD = "/dev/null"

    try:
        pid = os.fork()
        if pid != 0:
            sys.exit(0)
        else:
            os.setsid()
            # second fork to be sure the process is no process group leader
    #        pid = os.fork()
    #        if pid != 0:
    #            sys.exit(0)
    #        else:
            os.chdir(WORKDIR)
            os.umask(UMASK)

    except OSError as e:
        sys.stderr.write("fork  failed: {} ({})\n".format(e.errno, e.strerror))
    
    return os.getpid()

def run_as_daemon(config):
   
    if config['pid'] != None:
        pid_path = Path(config['pid'])
    else:
        pid_path = Path(PID_DEFAULT)

    if pid_path.is_file():
        sys.stderr.write("PID-File {} already exists, closing ..\n"
                .format(pid_path))
        sys.exit(-1)
    else:
        pid = daemonize()
        pid_path.write_text(str(pid))

    return pid

def get_file_from_url(url):
    try:
        response = urllib.request.urlopen(url)
        #print("returned status: {} - file size: {}".
        #    format(response.status,response.getheader("Content-Length")))
    except urllib.error.HTTPError as e:
        sys.stdout.write("HTTP-Error: maybe the station id is incorrect\n")
        sys.exit(-1)
    except urllib.error.URLError as e:
        sys.stdout.write("URL-Error: maybe internet connection is not yet established\n")
        raise
        
    return response.read()

def get_test_file():
    fb = open(station,"rb")
    f = fb.read()
    fb.close()
    return f

def check_date(wff):
    return wff[3][0:2]

def get_description(code,config):
    """Function returns the description of the present weather in german or english."""

    if code in [str(i) for i in range(32)]:
        if config['german']:
            return(PRESENT_WEATHER_CODES[code][1])
        else:
            return(PRESENT_WEATHER_CODES[code][2])
    else:
        return "NA"

def get_values(wff,config):
    return { 
                "time" : wff[3][1] +' (UTC)', # TODO: time offset for lokal time 
                "temp" : wff[3][9] + ' °C',
                "cloud-cover" : wff[3][2]+' %',
                "visibility" : wff[3][14]+' km',
                 "maxwind" : wff[3][18]+' km/h',
                 "wind-direction" : wff[3][22]+' °',
                 "precipitation" : wff[3][33]+' mm',
                 "weather" : get_description(wff[3][35],config),
                 "pressure" : wff[3][36]+' hPa',
                 "humidity" : wff[3][37]+' %',
            }

def write_file(vals, file_path):
    s = "{} T: {}  H: {}  W: {} ({})".format(
            vals["time"],
            vals["temp"],
            vals["humidity"],
            vals["maxwind"],
            vals["weather"]
            )

    with open(file_path, "w") as f:
        f.write(s)

def write_error(err_string, file_path):
    
    with open(file_path, "w") as f:
        f.write(err_string)

def parse_args():

    config = {
            'station_file'  : None,
            'pid'           : None,
            'daemon'        : False,
            'german'        : False
            }

    parser = argparse.ArgumentParser()

    parser.add_argument("-s", "--station", required=True,
            help="id of the weather station (mandatory)")
    parser.add_argument("-p", "--pid", help="path to the pid file (default {})"
            .format(PID_DEFAULT))
    parser.add_argument("-d", "--daemon", action="store_true",
            help="run as daemon (EXPERIMENTAL)")
    parser.add_argument("-g", "--german", action="store_true",
            help="display weather description in german")

    args = parser.parse_args()

    config['daemon'] = args.daemon
    config['german'] = args.german

    if args.station is None:
        sys.stderr.write("ERROR: station id needed\n")
        sys.exit(1)
    else:
        station = args.station
        if len(station) < 5:
            station +=  "_"
        config['station_file'] = station + "-BEOB.csv"

    return config

def read_weather(config):

    is_ok = True

    try:
        wf = get_file_from_url(DWD_POI_URL + config['station_file'])
        #wf = get_test_file()
    except urllib.error.URLError:
        wf = None
        is_ok = False

    if wf != None:
        wff = wf.decode("utf_8").splitlines()
        csv_reader = csv.reader(wff, delimiter=';')
        wff = [row for row in csv_reader]
        write_file(get_values(wff,config),FILE_PATH)
    else:
        write_error("service not available retrying ...",FILE_PATH)

    return is_ok

def main():
    
    config = parse_args()
    station_file_name = config['station_file']

    if os.name == 'posix':
        if os.getuid == 0 or os.geteuid == 0:
            sys.stderr.write("Do not run this script as root user!\n")
            sys.exit(1)

        if config['daemon']:
            run_as_daemon(config)
            while True:
                if read_weather(config):
                    sleep(1200)
                else:
                    while not read_weather(config):
                        sys.stderr.write("could not retrieve weather information,"
                                + " retrying ...\n")
                        sleep(60)
        else:
                read_weather(config)
    else:
        sys.stderr.write("Only for unix and compatible.\n")
        sys.exit(1)
    
if __name__ == "__main__":
    main()
